### DISCUSSION TITLE
Transforming the date and time

### DISCUSSION THREAD

**Discussion by micall2319** [nextgenhealthcare/connect#5950] [Q&A] [answered]
I have to create a channel in Mirth that reads HL7 files as the one attached from a local directory, and transforms them, and writes them to another channel. The transformation done to the HL7 file should be the following: For each timestamp in the message a capital T should be added between the date and time (20230418 13:46:48 -> 20230418T13:46:48)

Can anyone help me get started on this challenge? I am new to Mirth.

I have Mirth installed on my PC and I have a Word document with an HL7 message on it where I am supposed to add (or transform) a Capital T between the date and time on this HL7 message or template from one channel to another.

MSH|^~\&|SECA103|SECA|EPIC|EPIC|20230418 13:46:48||ORU^R01|667914f3-e8c1-4788-8b08-7b5e0d3a6506|P|2.4
PID|||P555^^^||Frank^Duke||19880505|M|||||||||||||||||
PV1||S|||||||||||||||||
OBR||||S|||20230418 13:46:48||||||||||||||||||F
OBX|1|NM|29463-7||82000|g||N|||F|||20230418 13:46:48||User1234||DeviceA
OBX|2|NM|8302-2||1,7900|m||N|||F|||20230418 13:46:48||User1234||DeviceA

**Comment by jonbartels** [comment, marked as answer]
This is a common question in Slack. The general rules that the community has found for dealing with dates and times and time zones in Mirth is:

1. Use `java.time` It is the new modern standard. https://docs.oracle.com/javase/8/docs/api/java/time/package-summary.html
2. Do not use `java.util.date` - It is out**dated** (HA HA) and `java.time` is better
3. Do not use the Mirth `DateUtil` class - it is based on `java.util.date`
4. Parse your input from a string to a time object using https://docs.oracle.com/javase/8/docs/api/java/time/format/DateTimeFormatter.html
5. Manipulate your time object as needed. Shift time zones, add minutes, etc.
6. Generate your output using https://docs.oracle.com/javase/8/docs/api/java/time/format/DateTimeFormatter.html 
7. When working with Zones use the full format and NEVER the daylight savings format. So US/Eastern or Europe/Berlin and not EDT/EST nor CDT/CET

Searching Mirth github and the forums for `java.time` should turn up some examples. Post your code here if you get stuck and need a review or more help.

If you make a code template - consider sharing it to https://github.com/nextgenhealthcare/connect-examples please!

**Reply by pacmano1** [reply]
@jonbartels , in this case the ask is just string manipulation, replacing the space with a "T".

@micall2319 , if you are completely new to mirth and not familiar with how it works at all, it's typically best to look here on github, the mirth forums, youtube, read the user guide,  etc for the "basics" on file readers, message transformations, etc.   Taking some formal training would be invaluable.  Then come back and post screenshots and code attempts with what you are struggling with.   

In general asking for a complete beginner walkthrough on how to build channels in Mirth doesn't get much of a response here since there is so much reference material already out there on the Internet.

chatGPT can be helpful here, but understanding HL7 and how mirth processes messages will go a long way to guide chatGPT to a correct answer. For example, chatGPT might choose to do a regex replace on the entire message via string manipulation vs iteraiting over specific segments and fields you wish to chage.
